test/amperity/gocd/secret/vault/plugin_api.clj

```diff
@@ -79,3 +79,45 @@
           (:content_type body))
       (is (:data body))
       (is (= status 200)))))
+
+
+(deftest secrets-lookup
+  (testing "Can look up secrets stored in vault given a well formed request"
+    (let [result (plugin/handle-request (mock-client) "go.cd.secrets.secrets-lookup"
+                                        {;; Eventual plan for configuration is to allow overrides to the user group and
+                                         ;;     re-authenticate the vault client with a new user group
+                                         :configuration {}
+                                         ;; The keys will likely be string in the http vault client instance, but this
+                                         ;;     is easier for testing.
+                                         :keys          [:batman :hulk :wonder-woman]})
+          body (:response-body result)
+          status (:response-code result)
+          _ (:response-headers result)]
+      (is (= [{:key :batman :value "Bruce Wayne"}
+              {:key :hulk :value "Bruce Banner"}
+              {:key :wonder-woman :value "Diana Prince"}]
+             body))
+      (is (= status 200))))
+  (testing "Fails cleanly when looking up secrets that don't exist"
+    (let [result (plugin/handle-request (mock-client) "go.cd.secrets.secrets-lookup"
+                                        {;; Eventual plan for configuration is to allow overrides to the user group and
+                                         ;;     re-authenticate the vault client with a new user group
+                                         :configuration {}
+                                         :keys          [:dr-who :jack-the-ripper]})
+          body (:response-body result)
+          status (:response-code result)
+          _ (:response-headers result)]
+      (is (= {:message "Unable to resolve key(s) [:dr-who :jack-the-ripper]"}
+             body))
+      (is (= status 404))))
+  (testing "Fails cleanly when other lookup error occurs"
+    (let [result (with-redefs [map (fn [_ _] (throw (Exception. "Mock Exception")))]
```

Comment:
Suggestion:

Though redefining `map` is clever, this could be done on `vault/read-secret`. That would make me less leery of repeating this pattern elsewhere or forgetting this is the case.

Reply:
I tried this on `vault/read-secret`. For some reason, the `with-redefs` doesn't work when I do that (the test just uses the actual `vault/read-secret` instead of the override). Definitely the better approach, but after trying for a while I just changed it to this. My guess is it has something to do with calling `vault/read-secret` as part of an anonymous function in the API endpoint. @pyncc maybe we can block out a time today/tomorrow to look at that together?

Reply:
This required some code changes so you should re-review!